### ISSUE TITLE
10-bit color support drastically reduced performance

### ISSUE THREAD

**Issue by chrisduerr** [Smithay/smithay#1023]
I've tested the time it takes to call `render` in my compositor and it went from ~1ms while starting up Firefox with a fixed website to ~10ms.

The bisect identified that one of the following commits is at fault:

```
594fb497ef37a5b7f7dc36cbbca77e8f3a65af65
c5b9baebb571fc2799b807a93062cc5c39f5e910
06389b115127b90c5607d2f8add97ec6c00247a0
99f605a51661100b75c4259ddcb1a8e9c7ae312c
57ceb42c72544ae5a1bfb6d7ba254fc73c62ab10
7196c3e0ae43c49a2163e8063703189d535e2551
318738bb84cbe82da3effd04a28cbee091542b7e
```

These don't build, so I could not bisect any further, but it looks like https://github.com/Smithay/smithay/pull/935 is likely at fault.

**Comment by chrisduerr** [comment]
I've been able to confirm that this line causes the performance issues: 

https://github.com/Smithay/smithay/blob/fa58fec5934a86c3a849869669cc1a9897850109/src/backend/renderer/gles/mod.rs#L587

**Comment by Drakulix** [comment]
I propose an API change, that makes it possible to query supported capabilities and to initialize the `GlesRenderer` with a set of requested capabilities (which fails if the requested ones aren't available).
If you are not using color transformations and you are never dealing with anything else, but 8-bit buffers, or your hardware is simply too slow, you can just drop the capability. But especially the types of buffers the compositor advertises or "too slow" implementations of set apis are nothing the renderer can figure out itself, so a user-driven approach seems to make sense here.

**Comment by chrisduerr** [comment]
That would work fine for me. I don't care manual configuration of this.

**Comment by chrisduerr** [comment]
@Drakulix I was hoping you'd have some time to implement this, but if this is not the case could you go into some more detail of what you'd expect the changes to look like so I can fix these performance issues?

**Comment by Drakulix** [comment]
> @Drakulix I was hoping you'd have some time to implement this, but if this is not the case could you go into some more detail of what you'd expect the changes to look like so I can fix these performance issues?

Sure thing, so the main thing to do here is splitting up `GlesRenderer::new` into two functions.

The current function is evaluating the feature-set to be used in this part of the code: https://github.com/Smithay/smithay/blob/master/src/backend/renderer/gles/mod.rs#L515-L603.

A new function (`evaluate_capabilities`? possibly a method of the `Capability`-type) would do essentially this and just return the capabilities array.

The `new` function would take a new `capabilities`-argument and compare the requested capabilities with the ones available. (So initializing the renderer in `anvil` would now evaluate the capabilities twice, but I don't think that is a huge issue. If one would want to avoid that, the new argument could be an `Option`, where `None` essentially means the previous behaviour of enabling everything possible.)

If the requested capabilities are not available, this has to throw a new error. Compositors like `catacomb` could then either hard-code as set of capabilities or dynamically drop the `ColorTransformation`-cap from the listed set for better performance.

This has a few remarks for the remaining renderers, that can wrap the GlesRenderer:
- The `GlowRenderer` needs to pass-through the capability arguments.
- The `MultiRenderer`s `EglGlesBackend` and `GbmGlesBackend` platforms initialize `GlesRenderer`s internally. They would probably need to take closure on construction and store that boxed, that returns a set of capabilities for maximum flexibility. Failing to request a given set of capabilities should likely not bring the whole `MultiRenderer` down, but simply log a warning and skip over the failed device in the platforms `enumerate` functions.

**Comment by cmeissl** [comment]
fixed by #1055
